typing/ctype.ml

```diff
@@ -182,6 +182,52 @@ let end_def () =
 let create_scope () =
   init_def (!current_level + 1);
   !current_level
+let wrap_def ?post f =
+  begin_def ();
+  let result = f () in
+  end_def ();
+  Option.iter (fun g -> g result) post;
+  result
+let wrap_def_process_if cond f ~proc =
+  if cond then begin_def ();
+  let result, l = f () in
+  if cond then begin
+    end_def ();
+    List.iter proc l;
+  end;
+  result
+let wrap_def_process f ~proc = wrap_def_process_if true f ~proc
+let wrap_def_if cond f ~post =
+  if cond then begin_def ();
+  let result = f () in
+  if cond then begin
+    end_def ();
+    post result;
+  end;
+  result
+let wrap_def_principal f ~post = wrap_def_if !Clflags.principal f ~post
+let wrap_def_process_principal f ~proc =
+  wrap_def_process_if !Clflags.principal f ~proc
+let wrap_init_def_if cond ~level f =
+  if cond then (begin_def (); init_def level);
+  let result = f () in
+  if cond then end_def ();
+  result
+let wrap_init_def ~level f = wrap_init_def_if true ~level f
+
+let wrap_class_def ?post f =
+  begin_class_def ();
+  let result = f () in
+  end_def ();
+  Option.iter (fun g -> g result) post;
+  result
+
+let wrap_raise_nongen_level f =
+  raise_nongen_level ();
+  let result = f () in
+  end_def ();
+  result
```

Comment:
These `warp_*` functions are unsafe if their `f` argument raises an exception. Could you make this extra clear by catching exceptions raised by `f` and failing with a `Failure` error in this case, or some other exception that is not intended to be caught by the callers (`Wrap_failure of exn`)? Otherwise I am worried that some code could use `wrap_*` inside a computation that raises an exception handled by the calling context, and not notice that the levels are completely off.

Reply:
Good question.
There are certainly exceptions we want to let go through, since this is the way typing errors were handled until now.
On the other hand, it might be reasonable to restore the original level in case of failure, since the semantics would be cleaner (only a bit, since level invariants are going to be probably brocken inside types anyway).

Reply:
All the wrapping functions now call
```ocaml
Misc.try_finally f ~always:end_def
```
This means that for instance in `check_counter_example_pat` there should be no need to keep and reinstate levels. We didn't change that yet.